rust/arrow/src/buffer.rs

```diff
@@ -861,49 +913,147 @@ impl MutableBuffer {
+    /// # Example
+    /// ```
+    /// # use arrow::buffer::MutableBuffer;
+    /// let mut buffer = MutableBuffer::new(0);
+    /// buffer.push(256u32);
+    /// assert_eq!(buffer.len(), 4) // u32 has 4 bytes
+    /// ```
+    #[inline]
+    pub fn push<T: ToByteSlice>(&mut self, item: T) {
+        let additional = std::mem::size_of::<T>();
+        self.reserve(additional);
         unsafe {
-            let dst = NonNull::new_unchecked(self.data.as_ptr().add(self.len));
-            let src = NonNull::new_unchecked(bytes.as_ptr() as *mut u8);
-            memory::memcpy(dst, src, bytes.len());
+            let dst = self.data.as_ptr().add(self.len) as *mut T;
+            std::ptr::write(dst, item);
         }
-        self.len = new_len;
+        self.len += additional;
+    }
+
+    /// Extends the buffer by `additional` bytes equal to `0u8`, incrementing its capacity if needed.
+    #[inline]
+    pub fn extend_zeros(&mut self, additional: usize) {
+        self.resize(self.len + additional, 0);
     }
+}
 
-    /// Extends the buffer by `len` with all bytes equal to `0u8`, incrementing its capacity if needed.
-    pub fn extend(&mut self, len: usize) {
-        let remaining_capacity = self.capacity - self.len;
-        if len > remaining_capacity {
-            self.reserve(self.len + len);
+impl<A: ArrowNativeType> Extend<A> for MutableBuffer {
+    /// Extends [`MutableBuffer`] from an [`Iterator`]. If the iterator is an [`ExactSizeIterator`],
+    /// use [`extend_from_exact_iter`] as it is faster.
+    #[inline]
+    fn extend<T: IntoIterator<Item = A>>(&mut self, iter: T) {
+        let iterator = iter.into_iter();
+        self.extend_from_iter(iterator)
+    }
+}
+
+// This implementation is required for two reasons:
+// 1. In stable rust, there is no trait specialization and therefore
+//    we can't specialize `extend_from_iter` for `ExactSizeIterator` like `Vec` does.
+// 2. Extend for `ExactSizeIterator` is at least 2x faster (different scaling).
+// Unfortunately, there is no "FromExactSizeIterator", only `FromIterator`,
+// which means that `collect` is slow and we should avoid it to build a Buffer. This should
+impl MutableBuffer {
+    /// This method is private because the public interface is [`extend`]
+    /// It is here because it is similar to `extend_from_exact_iter`.
+    fn extend_from_iter<T: ArrowNativeType, I: Iterator<Item = T>>(
+        &mut self,
+        mut iterator: I,
+    ) {
+        let size = std::mem::size_of::<T>();
+
+        let (lower, _) = iterator.size_hint();
+
+        self.reserve(lower * size);
+        let mut dst = unsafe { self.data.as_ptr().add(self.len) as *mut T };
+
+        while let Some(item) = iterator.next() {
+            if self.len >= self.capacity() {
+                let (lower, _) = iterator.size_hint();
+                let additional = (lower + 1) * size;
+                self.reserve_at_least(self.len + additional);
```

Comment:
I am sorry, I did not understand your question.

This is being called here because iterators do not have a reliable `size_hint` (thus the `+ 1`). This code is spiritually equal to `extend_desugared`  in the standard lib.